src/kvstore.h

```diff
@@ -0,0 +1,81 @@
+#ifndef DICTARRAY_H_
+#define DICTARRAY_H_
+
+#include "dict.h"
+#include "adlist.h"
+
+typedef struct {
+    list *rehashing;                       /* List of dictionaries in this kvstore that are currently rehashing. */
+    int resize_cursor;                     /* Cron job uses this cursor to gradually resize dictionaries. */
+    int non_empty_dicts;                   /* The number of non-empty dicts. */
+    unsigned long long key_count;          /* Total number of keys in this kvstore. */
+    unsigned long long bucket_count;       /* Total number of buckets in this kvstore across dictionaries. */
+    unsigned long long *dict_size_index;   /* Binary indexed tree (BIT) that describes cumulative key frequencies up until given dict-index. */
+} kvstoreState;
+
+typedef struct {
```

Comment:
Can we make this an opaque pointer, i.e. only define it inside the .c file, so external users go through the APIs and don't try to access anything. Same with the iterator I suppose.

Reply:
i'd love that, but it won't make sense to have everything opaques and still give direct dict access

relates to https://github.com/redis/redis/pull/12822#discussion_r1464132396